When playing sound through an AudioContext, stale audio buffer data plays after seek

jyavenard opened the pull request:
#### 80b43161244b75fb66a86a352c81d5b15e7045a2
<pre>
When playing sound through an AudioContext, stale audio buffer data plays after seek
<a href="https://bugs.webkit.org/show_bug.cgi?id=288879">https://bugs.webkit.org/show_bug.cgi?id=288879</a>
<a href="https://rdar.apple.com/146057507">rdar://146057507</a>

Reviewed by Youenn Fablet.

Prior the move of the AudioSourceProvider to the GPU process,
AudioSourceProvider::provideInput would be regularly called to get
the next audio data to be provided back to an AudioContext.
AudioSourceProviderAVFObjC::provideInput would detect if a flush or seek
occurred and would then discard any stale content before returning new data.

Since 230539@main however, the audio produced by the media player in the GPU
process is immediately sent to the WebContent process and stored in the
AudioSampleDataSource&apos;s ring buffer ready to be retrieved by the AudioContext,
in effect deprecating AudioSourceProviderAVFObjC::provideInput (and making
it dead code).
As such, when we reset or seek the media player, any data set to become stale
has already been queued in the WP and will be pulled next.

To get around this issue, we sent to the content process the information
that earlier data received is now stale and needs to be flushed.
When the AudioSampleDataSource::pullSamples is next called, any audio data
still pending in the AudioSampleDataSource received earlier than the data
following the request for flushing will be dropped.

Manually verified using provided test site.

* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.mm:
(WebCore::AudioSampleDataSource::pushSamplesInternal):
(WebCore::AudioSampleDataSource::pushSamples):
(WebCore::AudioSampleDataSource::pullSamples):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::process):
* Source/WebCore/platform/mediastream/mac/WebAudioSourceProviderCocoa.h:
* Source/WebCore/platform/mediastream/mac/WebAudioSourceProviderCocoa.mm:
(WebCore::WebAudioSourceProviderCocoa::receivedNewAudioSamples):
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.cpp:
(WebKit::RemoteAudioSourceProviderProxy::create):
(WebKit::RemoteAudioSourceProviderProxy::newAudioSamples):
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProvider.cpp:
(WebKit::RemoteAudioSourceProvider::audioSamplesAvailable):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProvider.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.cpp:
(WebKit::RemoteAudioSourceProviderManager::audioSamplesAvailable):
(WebKit::RemoteAudioSourceProviderManager::RemoteAudio::audioSamplesAvailable):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.messages.in:

Canonical link: <a href="https://commits.webkit.org/291763@main">https://commits.webkit.org/291763@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50aa763505fd43fed8eb125d313d1cbde7626cc4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98967 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44486 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96009 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13842 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21975 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71693 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29043 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10269 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52032 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9956 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2522 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43802 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80212 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/2604 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101006 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21011 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80703 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80846 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80062 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19929 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24609 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1969 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14168 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20995 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26173 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20682 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24142 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22423 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.mm:
```diff
@@ -207,7 +210,12 @@
         return false;
     }
 
+    uint64_t seekTo = std::exchange(m_seekTo, NoSeek);
```
push and pull may be done in different threads, we probably want to have `m_seekTo` as an atomic.

jyavenard replied:
> push and pull may be done in different threads, we probably want to have `m_seekTo` as an atomic.

The code as it is existing isn't thread-safe. We have several non-atomic member variables being accessed in both the pull and push method: `m_waitToStartForPushCount` , `m_muted` or even the `m_scratchBuffer` ; if push and pull were concurrent there would be serious trouble.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.mm:
```diff
@@ -207,7 +210,12 @@
 
+    uint64_t seekTo = std::exchange(m_seekTo, NoSeek);
+    if (seekTo != NoSeek)
+        m_readCount = seekTo;
```
Do we need m_readCount global variable?

jyavenard replied:
yes, while a push will typically be with like 4096 frames at a time, from what I've seen pulls are small (128) and this is why we typically have stale data in wait.

So we want to keep track on where we will start from during the next few pulls. The index of the reads is monotonically increasing.

m_readCount indicates that all date prior that index are stale.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.mm:
```diff
@@ -171,6 +171,9 @@
 
+    if (needsFlush == NeedsFlush::Yes)
+        m_seekTo = ringBufferIndexToWrite;
+
```
I wonder what happens if this pull code is called when m_seekTo is set but storage in the ring buffer is not done.
In that case, the pull will fail but will the next pull start where it should?

jyavenard replied:
that's the job of m_readCount.

If the data hasn't been written yet in the ring buffer, then in the pull method `m_ringBuffer->getFetchTimeBounds();` will show that the data isn't ready yet.

youennf reviewed (commented):
Approach LGTM in general, some questions below before r+

youennf reviewed (approved):
r=me, please make m_seekTo atomic and add a comment that m_readCount is accessed only in the thread that is puling.

jyavenard committed:
When playing sound through an AudioContext, stale audio buffer data plays after seek
https://bugs.webkit.org/show_bug.cgi?id=288879
rdar://146057507

Reviewed by Youenn Fablet.

Prior the move of the AudioSourceProvider to the GPU process,
AudioSourceProvider::provideInput would be regularly called to get
the next audio data to be provided back to an AudioContext.
AudioSourceProviderAVFObjC::provideInput would detect if a flush or seek
occurred and would then discard any stale content before returning new data.

Since 230539@main however, the audio produced by the media player in the GPU
process is immediately sent to the WebContent process and stored in the
AudioSampleDataSource's ring buffer ready to be retrieved by the AudioContext,
in effect deprecating AudioSourceProviderAVFObjC::provideInput (and making
it dead code).
As such, when we reset or seek the media player, any data set to become stale
has already been queued in the WP and will be pulled next.

To get around this issue, we sent to the content process the information
that earlier data received is now stale and needs to be flushed.
When the AudioSampleDataSource::pullSamples is next called, any audio data
still pending in the AudioSampleDataSource received earlier than the data
following the request for flushing will be dropped.

Manually verified using provided test site.

* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.mm:
(WebCore::AudioSampleDataSource::pushSamplesInternal):
(WebCore::AudioSampleDataSource::pushSamples):
(WebCore::AudioSampleDataSource::pullSamples):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::process):
* Source/WebCore/platform/mediastream/mac/WebAudioSourceProviderCocoa.h:
* Source/WebCore/platform/mediastream/mac/WebAudioSourceProviderCocoa.mm:
(WebCore::WebAudioSourceProviderCocoa::receivedNewAudioSamples):
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.cpp:
(WebKit::RemoteAudioSourceProviderProxy::create):
(WebKit::RemoteAudioSourceProviderProxy::newAudioSamples):
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProvider.cpp:
(WebKit::RemoteAudioSourceProvider::audioSamplesAvailable):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProvider.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.cpp:
(WebKit::RemoteAudioSourceProviderManager::audioSamplesAvailable):
(WebKit::RemoteAudioSourceProviderManager::RemoteAudio::audioSamplesAvailable):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.messages.in:

Canonical link: https://commits.webkit.org/291763@main
